src/main/scala/mesosphere/marathon/core/storage/impl/cache/LazyCachingPersistenceStore.scala

```diff
@@ -0,0 +1,142 @@
+package mesosphere.marathon.core.storage.impl.cache
+
+import java.time.OffsetDateTime
+
+import akka.http.scaladsl.marshalling.Marshaller
+import akka.http.scaladsl.unmarshalling.Unmarshaller
+import akka.stream.ActorMaterializer
+import akka.stream.scaladsl.{ Keep, Sink, Source }
+import akka.{ Done, NotUsed }
+import com.typesafe.scalalogging.StrictLogging
+import mesosphere.marathon.core.storage.impl.BasePersistenceStore
+import mesosphere.marathon.core.storage.{ CategorizedKey, IdResolver, PersistenceStore }
+import mesosphere.marathon.util.RwLock
+
+import scala.async.Async.{ async, await }
+import scala.collection.concurrent.TrieMap
+import scala.collection.immutable.Seq
+import scala.concurrent.{ ExecutionContext, Future }
+
+/**
+  * A Write Ahead Cache of another persistence store that lazily loads values into the cache.
+  *
+  * @param store The store to cache
+  * @param mat a materializer for Akka Streaming
+  * @param ctx The execution context for future chaining.
+  * @tparam K The persistence store's primary key type
+  * @tparam Serialized The serialized format for the persistence store.
+  */
+class LazyCachingPersistenceStore[K, Category, Serialized](
+    store: BasePersistenceStore[K, Category, Serialized])(implicit
+  mat: ActorMaterializer,
+    ctx: ExecutionContext) extends PersistenceStore[K, Category, Serialized] with StrictLogging {
+
+  private[storage] val idCache = RwLock(TrieMap[Category, Seq[Any]]())
```

Comment:
Nit: to distinguish type from value I think it easier to read `TrieMap.empty[A,B]`